BeanFactory support with autoconversion

zstd opened the pull request:
Hello.

Improved support for BeanConfig mentioned in pull request #107 
(https://github.com/typesafehub/config/pull/107)

Added support for automatic type conversions plus some tests for those features.

Ready for questions and comments.

benmccann commented:
awesome!

@zstd the build is failing on this PR. Looks like you're using Java 7 and this library is supporting Java 6 still (see https://github.com/typesafehub/config/blob/master/.travis.yml#L18)

    [error] /home/travis/build/typesafehub/config/config/src/main/java/com/typesafe/config/impl/BeanFactory.java:27: illegal start of type
    [error]     private static final Map<TimeUnit,List<String>> TIME_UNITS = new HashMap<>();

benmccann commented on config/src/main/java/com/typesafe/config/impl/BeanFactory.java:
```diff
@@ -0,0 +1,150 @@
+
+    private static final Map<TimeUnit,List<String>> TIME_UNITS = new HashMap<>();
+    static {
+        TIME_UNITS.put(TimeUnit.NANOSECONDS, Arrays.asList("ns", "nano", "nanos", "nanosecond", "nanoseconds"));
```
Are these all valid? It doesn't look like what is being used elsewhere in the code. See:
https://github.com/typesafehub/config/blob/master/config/src/main/java/com/typesafe/config/impl/SimpleConfig.java#L522

zstd replied:
I've taken form HOCON spec
https://github.com/typesafehub/config/blob/master/HOCON.md#duration-format

benmccann commented:
It looks like there are other automatic type conversions not handled here. E.g. what about bytes?
https://github.com/typesafehub/config/blob/master/HOCON.md

Ideally I think there might be a bit more code reuse with some of the existing code

havocp commented on config/src/main/java/com/typesafe/config/impl/BeanFactory.java:
```diff
@@ -0,0 +1,160 @@
+/**
+ * Sent as pull request to config project.
+ * https://github.com/typesafehub/config/pull/107 (original)
+ * https://github.com/typesafehub/config/pull/249 (automatic conversion support)
```
this comment will need to come out before merge

zstd replied:
Sorry, this is my first pull request. Can you explain what wrong with this lines?

benmccann replied:
They just add noise to the source code without adding any useful information

havocp commented on config/src/main/java/com/typesafe/config/impl/BeanFactory.java:
```diff
@@ -0,0 +1,160 @@
+ * https://github.com/typesafehub/config/pull/107 (original)
+ * https://github.com/typesafehub/config/pull/249 (automatic conversion support)
+ */
+public class BeanFactory {
```
`ConfigBeanFactory` probably matches the other types in the package better, though I don't really know why we Config-prefixed everything I guess at this point we should go with it.

havocp commented on config/src/main/java/com/typesafe/config/impl/BeanFactory.java:
```diff
@@ -0,0 +1,160 @@
+            } else if (isByteValue(rawVal)) {
+                return config.getBytes(configPropName);
+            }
+            return config.getLong(configPropName);
```
This heuristic scares me a bit, so I'm trying to think of alternatives; maybe an annotation? I'm not sure what the annotation looks like though because we don't really have an enum right now that contains things like `DURATION` or `BYTES` so we'd have to add a thing like that.

benmccann replied:
really I think the best way to do this would to have it be a java.time.Duration instead of a long. the real problem here is that we're overloading long and using it not a in typesafe manner. of course to use java.time.Duration we'd have to support Java 8. i think it wouldn't be a bad call to support Java 8 because Java 7 is EOL (end of life) in April. Play is currently discussing releasing Play 2.4 only for Java 8 for that reason. See https://groups.google.com/forum/#!topic/play-framework-dev/fte3TC4rYNw

havocp replied:
that would solve duration but not bytes. I guess we could go Java 8 if Akka and play both do, though. should probably be a separate discussion / PR.

havocp commented on config/src/main/java/com/typesafe/config/impl/BeanFactory.java:
```diff
@@ -0,0 +1,160 @@
+    }
+
+
+    static TimeUnit getTimeUnit(String rawVal) {
```
here you could probably reuse the existing parseDuration https://github.com/typesafehub/config/commit/83ceb90a895107b1f261e2a9af319d4c1cf4194c#diff-6f3de88fcbbb2a57e42de159f09350d6L497

havocp replied:
here the existing parseDuration can likely be reused: https://github.com/typesafehub/config/blob/master/config/src/main/java/com/typesafe/config/impl/SimpleConfig.java#L497

zstd replied:
My idea of not using parseDuration was that in case of BeanConfig I try to perform autoconversion of duration (and bytes too) only if marker representing time (size) unit (i.e. 's','ms','seconds' etc) is present in rawString. I have no meta information how to treat string --> long conversion so I inspect string for some 'markers' for duration or size and then perform conversion. Of course if we can have come meta-info annotations then it will be possible to reuse parseDuratation. Another way may be to override parseDuration with some flag indicating that we should parse only string containing markers.

benmccann replied:
I think you'll still need isDuration and you'll have to do the isDuration check before calling parseDuration, but it doesn't seem like you need this method getTimeUnit

zstd replied:
You suppose to return always duration is nano-seconds? 
I see the profit of moving isDuration to Config but in that case I will add new method to Config interface. Is that acceptable? If yes, what tests can I check to be sure that I am not broking something.

benmccann replied:
You can run `sbt test` to run the tests and ensure nothing is breaking

havocp commented on config/src/main/java/com/typesafe/config/impl/BeanFactory.java:
```diff
@@ -0,0 +1,160 @@
+                    throw new ConfigException.Generic(
+                            "Could not find " + beanProp.getName() + " from " + clazz.getName() + " in config.");
+                }
+                if (configValue instanceof Map) {
```
`instanceof ConfigObject` would be better here I think

zstd replied:
Actually current implementation here deals only with 'raw' view so no ConfigObjects here.

havocp commented on config/src/main/java/com/typesafe/config/impl/BeanFactory.java:
```diff
@@ -0,0 +1,160 @@
+            return config.getInt(configPropName);
+        } else if (parameterClass == Double.class || parameterClass == double.class) {
+            return config.getDouble(configPropName);
+        } else if (parameterClass == Long.class || parameterClass == long.class) {
```
trying to think what types we might want to add here; we could do the byte, char, short, and List maybe? Also how about ConfigValue/ConfigObject/ConfigList (which would not do any conversion but give you the raw thing)?

havocp commented on config/src/main/java/com/typesafe/config/impl/BeanFactory.java:
```diff
@@ -0,0 +1,160 @@
+            }
+            return bean;
+        } catch (IntrospectionException e) {
+            throw new ConfigException.Generic("Could not resolve a string method name.", e);
```
is it possible to include the name we couldn't resolve in this message?

Is this exception always a bug in the program? If so we could use `ConfigException.BugOrBroken`

zstd replied:
This exception is raised in case when there is a property in ConfigClass that is not present in config file. I.e. we assume that config should set all class properties.
May be it would be better to add some ParsingOptions (for example STRICT(for above behaviour whan we require all props to be present and NON_STRICT when we simply skip non set properties))?

benmccann replied:
I don't think we need parsing options necessarily, especially in this first version. But it would be nice to say which method name we couldn't find

zstd replied:
Current error message is pretty obvious, I think:

Could not find property **'propNotListedInConfig'** from class **'beanconfig.NoFoundPropBeanConfig'** in config.

benmccann replied:
@zstd that's not what the error message we're talking about here says at all. it says "Could not resolve a string method name."

zstd replied:
Ok. I think introspection exception is more about loading class meta info so I refactored it a little.

havocp commented:
Thanks for taking this on! By the way I'm experimenting with a chat room: https://gitter.im/typesafehub/config 
if you want to discuss anything in a more interactive format

benmccann commented on config/src/main/java/com/typesafe/config/impl/BeanFactory.java:
```diff
@@ -0,0 +1,160 @@
+        return config.getAnyRef(configPropName);
+    }
+
+    private static boolean isDuration(String rawVal) {
```
i think the nice thing to do would be to put this in SimpleConfig and make it public. this should go along well with parseDuration

zstd replied:
Can you explain what actually can be put into SimpleConfig and what profit we will have from that?

benmccann replied:
SimpleConfig has some lines which contain the time units like:

    } else if (unitString.equals("us") || unitString.equals("micros") || unitString.equals("microseconds")) {

It would be nice to use TIME_UNITS for that as well so that there's only one list of time units we need to maintain instead of two

havocp commented:
This is looking quite good. My hesitation on merging this is mostly that it doesn't work the same way as config normally does on bytes and duration - ideally, we would simply call getDuration and not try to guess based on the config value whether it's a duration. The spirit of the library is that we always supply the type in code, rather than including type information in the config file itself.

if we had the Java 8 duration type, then we could skip the heuristics for duration, but not for bytes.

So I don't see a way around an annotation, when the field is a long. not sure of the best name... `@configValue(ConfigHint.DURATION)` ? When using the Java 8 duration type (which we could add support for later after merging java 8 support) we could of course allow people to omit it. But we'd still need the annotation for bytes.

benmccann commented:
@havocp another potential solution would be to build our own `Memory` type. that might actually be rather helpful to have as a user because we could then easily get the config value in any unit we want e.g. if we want to display to the user in MB or MiB and then pass to some api as bytes, then we don't need to duplicate the conversion logic that the config library already has again in our own code

havocp commented:
That could work, I'm open to that too. So use Java 8 duration and add a MemorySize type amd use that, no heuristics?

We will block on debugging that java 8 stuff though.

benmccann commented:
Yeah, that's kind of what I was thinking

zstd commented:
As the general comment:
can we list all fixes and improvements needed for this pull request to be ready for merging?

havocp commented:
@zstd I'm sorry there have been a lot of comments and some conflicting ones. If there are conflicting ideas, please feel welcome to ask for clarification. Also, I am always happy to fix small issues myself before merging, so you don't necessarily have to fix everything, but it may take some time for me to do it.

That said, if you (or someone else) wants to pick it up, I think there are two ways we could go forward.

One, we could first get Java 8 support merged (https://github.com/typesafehub/config/pull/252 ) which is waiting on debugging exactly why the tests pass on 6 but not on 8. I think the fixes to the tests would ideally be their own PR separate from the Java 8 PR. After Java 8 is merged we could come back to this PR and use the Java 8 duration type and add a `com.typesafe.config.MemorySize` type.

Two, we could remove duration and bytes support from this PR for the time being, and merge only the basic primitive type functionality. And then we could add duration and/or MemorySize at a later time.
That would make this PR smaller and it would be some work to split it up, but then we wouldn't be waiting for Java 8. In the Duration/MemorySize PR, we could also add getters to `Config` itself for those.

Once that's sorted out I don't think there are any large changes. The tests look good and the basic feature looks good.

The reason to use Duration and MemorySize is that having a field of type Foo in a bean will be exactly equivalent to calling `config.getFoo`, while if we have to heuristically guess types, beans behave differently from regular getters. For example, if you "getLong" on the value "10s" right now that would be an error I think (or maybe ignore the units, or something, but it won't be 10,000 as far as I know). The spec itself (HOCON.md) does not include "type hints" functionality, because the intent is that the "config getter" always provides the expected type. So if we do a bean feature that doesn't allow the "config getter" to provide the type it expects, we have a tension between the bean feature and the spec.

I did notice just now that there's no public API in the PR as it stands; everything is in impl, which is the private implementation package. I believe `ConfigBeanFactory.create` is intended to be public so shouldn't be in the impl package. Some of the static methods in ConfigBeanFactory like toCamelCase should probably be either `private` or in the impl package.

I think we also need a doc comment on the single public method (`create`) explaining how this feature works.

Thank you!

zstd commented:
Hello.

1) Added documentation, example of calling the bean. Moved from impl package.
2) Regarding the duration/memory functionality: can you see the tests, especially ConfigBeanFactoryTest.testCreateDuration/testCreateBytes? I think they deal pretty well for most of the cases right now. In case you think we should abandon this functionality right now anyway let me know, I'll remove it and left only autoconversion for numbers/booleans etc.

Thank you.

havocp commented:
Apologies for the delay. I sorted the Java 8 stuff I think, see #255 , and I added a ConfigMemorySize #256. I think I'd probably do a small PR very similar to #256 to add duration getters to Config. Then with all that on master we can rebase this PR and rely on all bean fields having useful types.

zstd commented:
Great. Let me know if something needs to be improved.

havocp commented:
#257 is the duration support.

benmccann commented:
@zstd the changes for duration and ConfigMemorySize have been merged, so you can rebase this PR now

@havocp thanks for all the help with those changes!! i'm really excited about this

havocp commented:
I've rebased this one already; branch coming up

havocp commented:
Rebased and moved to https://github.com/typesafehub/config/pull/264 . There are a couple more things I might like to change but feel free to comment and find bugs.
